### PR TITLE
lapack: add Dlanhs and its test

### PR DESCRIPTION
**Pull request by soypat**
Part of #1651.

Used dlange to validate.

### CODE REVIEW

File: lapack/testlapack/dlanhs.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	"gonum.org/v1/gonum/lapack"
+)
+
+type Dlanhser interface {
+	Dlanhs(norm lapack.MatrixNorm, n int, a []float64, lda int, work []float64) (resultNorm float64)
+}
+
+func DlanhsTest(t *testing.T, impl Dlanhser) {
+	const tol = 1e-8
+	work := make([]float64, 9*9)
```
Comment (vladimir-ch) [inline lapack/testlapack/dlanhs.go:21]: Please make this the minimum required length and fill it with garbage before every call to Dlanhs.

File: lapack/testlapack/dlanhs.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	rnd := rand.New(rand.NewSource(1))
+	for _, n := range []int{1, 2, 4, 9} {
+		for _, lda := range []int{n, n + 5} {
+			a := randomHessenberg(n, lda, rnd)
+			for _, norm := range []lapack.MatrixNorm{lapack.MaxAbs, lapack.MaxRowSum, lapack.MaxColumnSum, lapack.Frobenius} {
+				expect := dlange(norm, a.Rows, a.Cols, a.Data, a.Stride)
```
Comment (vladimir-ch) [inline lapack/testlapack/dlanhs.go:27]: Conventionally we call such variables `want`.

File: lapack/gonum/dlanhs.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import (
+	"math"
+
+	"gonum.org/v1/gonum/lapack"
+)
+
+// Dlanhs  returns the value of the one norm,  or the Frobenius norm, or
+// the  infinity norm,  or the  element of  largest absolute value  of a
+// Hessenberg matrix A.
+func (impl Implementation) Dlanhs(norm lapack.MatrixNorm, n int, a []float64, lda int, work []float64) (resultNorm float64) {
```
Comment (vladimir-ch) [inline lapack/gonum/dlanhs.go:16]: I wouldn't name the return, it's clear the float64 must be the norm of A. Also in the Dlanhser interface in the test.

File: lapack/testlapack/dlanhs.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+// license that can be found in the LICENSE file.
+
+package testlapack
+
+import (
+	"math"
+	"testing"
+
+	"golang.org/x/exp/rand"
+	"gonum.org/v1/gonum/lapack"
```
Comment (vladimir-ch) [inline lapack/testlapack/dlanhs.go:12]: Blank line between us and them.

File: lapack/gonum/dlanhs.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	if n == 0 {
+		return 0 // Early return.
+	}
+	var value, sum float64
+	switch norm {
+	default:
+		panic(badNorm)
+	case lapack.MaxAbs:
+		// Find max(abs(A(i,j))).
+		for j := 0; j < n; j++ {
```
[the remark this answers is not in the document]
Comment (soypat) [reply lapack/gonum/dlanhs.go:34]: How would you switch the loops?

[the remark this answers is not in the document]
Comment (kortschak) [reply lapack/gonum/dlanhs.go:36]: Something like this (untested).

```
		for i := 0; i < n; i++ {
			for _, v := range a[i*lda:i*lda+i+1] {
				value = math.Max(value, math.Abs(v))
			}
		}
```

[the remark this answers is not in the document]
Comment (vladimir-ch) [reply lapack/gonum/dlanhs.go:36]: Or use another Dlang? routine as a template, for example Dlangb.

[the remark this answers is not in the document]
Comment (soypat) [reply lapack/gonum/dlanhs.go]: Alright, suggestion applied!

File: lapack/gonum/dlanhs.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+		}
+		for j := 0; j < n; j++ {
+			value = math.Max(value, work[j])
+		}
+	case lapack.MaxRowSum:
+		for i := 0; i < n; i++ {
+			sum := 0.0
+			for j := max(0, i-1); j < n; j++ {
+				sum += math.Abs(a[i*lda+j])
+			}
```
Comment (kortschak) [inline lapack/gonum/dlanhs.go:58]: This can be Dasum over `a[max(0, i-1), n]` inc 1.

File: lapack/gonum/dlanhs.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+		return 0 // Early return.
+	}
+
+	var value float64
+	switch norm {
+	default:
+		panic(badNorm)
+	case lapack.MaxAbs:
+		for i := 0; i < n; i++ {
+			for j := max(0, i-1); j < n; j++ {
```
Comment (kortschak) [inline lapack/gonum/dlanhs.go:37]: Since this is inc 1, the range over the slice is probably going to be faster.

File: lapack/gonum/dlanhs.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+				value = math.Max(value, math.Abs(a[i*lda+j]))
+			}
+		}
+	case lapack.MaxColumnSum:
+		for i := 0; i < n; i++ {
+			work[i] = 0
+		}
+		for i := 0; i < n; i++ {
+			for j := max(0, i-1); j < n; j++ {
```
Comment (kortschak) [inline lapack/gonum/dlanhs.go:46]: Same here.

File: lapack/gonum/dlanhs.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+				work[j] += math.Abs(a[i*lda+j])
+			}
+		}
+		for j := 0; j < n; j++ {
```
Comment (kortschak) [inline lapack/gonum/dlanhs.go:50]: And here.

File: lapack/gonum/dlanhs.go (hunk 8 above)
Comment (soypat) [reply lapack/gonum/dlanhs.go:49]: but we'd have to to an operation to access work which i feel would not be as clear to the reader.

Comment (kortschak) [reply lapack/gonum/dlanhs.go:49]: You are right.

**Comment by soypat** [comment]
Are there blockers left? i.e. code coverage

**Comment by soypat** [comment]
@vladimir-ch You were right, there were many things left to fix. Hopefully got them all in last commit.

**Review by vladimir-ch — APPROVED** [review]

**Comment by vladimir-ch** [comment]
I'll resolve the remaining pending changes in another PR.

**Comment by kortschak** [comment]
Ooops. This should have been squashed.

**Comment by vladimir-ch** [comment]
Argh, I'm sorry.

**Pull request merged**